codex/blockexchange/engine/engine.nim

```diff
@@ -100,124 +93,147 @@ type
-    # prioritization
+proc downloadInternal(
+    self: BlockExcEngine, address: BlockAddress
+) {.async: (raises: []).} =
+  logScope:
+    address = address
 
-    # drop unresponsive peer
-    await b.network.switch.disconnect(peerId)
-    b.discovery.queueFindBlocksReq(@[address.cidOrTreeCid])
-
-proc pickPseudoRandom(
-    address: BlockAddress, peers: seq[BlockExcPeerCtx]
-): BlockExcPeerCtx =
-  return peers[hash(address) mod peers.len]
-
-proc requestBlock*(
-    b: BlockExcEngine, address: BlockAddress
-): Future[?!Block] {.async.} =
-  let blockFuture = b.pendingBlocks.getWantHandle(address, b.blockFetchTimeout)
+  let handle = self.pendingBlocks.getWantHandle(address)
+  trace "Downloading block"
+  try:
+    while address in self.pendingBlocks:
+      logScope:
+        retries = self.pendingBlocks.retries(address)
+        interval = self.pendingBlocks.retryInterval
+
+      if self.pendingBlocks.retriesExhausted(address):
+        trace "Error retries exhausted"
+        handle.fail(newException(RetriesExhaustedError, "Error retries exhausted"))
+        break
+
+      trace "Running retry handle"
+      let peers = self.peers.getPeersForBlock(address)
+      logScope:
+        peersWith = peers.with.len
+        peersWithout = peers.without.len
+
+      trace "Peers for block"
+      if peers.with.len > 0:
+        self.pendingBlocks.setInFlight(address, true)
+        await self.sendWantBlock(@[address], peers.with.randomPeer)
+      else:
+        self.pendingBlocks.setInFlight(address, false)
+        if peers.without.len > 0:
+          await self.sendWantHave(@[address], peers.without)
+        self.discovery.queueFindBlocksReq(@[address.cidOrTreeCid])
 
-  if not b.pendingBlocks.isInFlight(address):
-    let peers = b.peers.getPeersForBlock(address)
+      await (handle or sleepAsync(self.pendingBlocks.retryInterval))
+      self.pendingBlocks.decRetries(address)
 
-    if peers.with.len == 0:
-      b.discovery.queueFindBlocksReq(@[address.cidOrTreeCid])
-    else:
-      let selected = pickPseudoRandom(address, peers.with)
-      asyncSpawn b.monitorBlockHandle(blockFuture, address, selected.id)
-      b.pendingBlocks.setInFlight(address)
-      await b.sendWantBlock(@[address], selected)
+      if handle.finished:
+        trace "Handle for block finished", failed = handle.failed
+        break
+  except CancelledError as exc:
+    trace "Block download cancelled"
+    if not handle.finished:
+      await handle.cancelAndWait()
+  except CatchableError as exc:
+    warn "Error downloadloading block", exc = exc.msg
+    if not handle.finished:
+      handle.fail(exc)
+  finally:
+    self.pendingBlocks.setInFlight(address, false)
 
-    await b.sendWantHave(@[address], peers.without)
+proc requestBlock*(
+    self: BlockExcEngine, address: BlockAddress
+): Future[?!Block] {.async: (raises: [CancelledError]).} =
+  if address notin self.pendingBlocks:
+    self.trackedFutures.track(self.downloadInternal(address))
```

Comment:
So, what are we really tracking here in:

```nim
self.trackedFutures.track(self.downloadInternal(address))
```

 Which future will the `downloadInternal` return? If I read it well it will be the result of `await (handle or sleepAsync(self.pendingBlocks.retryInterval))`, which returns a new future which will complete once either `handle` or `sleepAsync` complete, which could well be `sleepAsync`. I think we always want to track `handle`.

Reply:
Also notice that having `let fut = await (fut1 or fut2)`, if `fut` is cancelled, `fut1` and `fut2` futures WILL NOT BE cancelled.

Reply:
well, in my reasoning, I ignored the fact that `downloadInternal` has a `while` loop. Thus, `downloadInternal` will return another Future I guess, which we are effectively tracking, but it still looks to me that `handle` will not be cancelled because of the "or" thing - I tried to test that with some examples, and it looks like indeed this is the case.